### ISSUE TITLE
Containers with only non-default constructor no longer work

### ISSUE THREAD

**Issue by sagamusix** [ThePhD/sol2#725]
I haven't paid attention to sol2 development for a while, but now I'm getting back into it and so I just tried to upgrade from 2.19.something to 2.20.3. Not entirely unexpected, but now my code does not compile anymore...
Since it's been a long while, I might have missed something important, but I cannot find anything it in the release notes. The problem appears to be that my container objects do not have a default constructor, and I cannot easily add one because they contain references to the objects they track. Now `getter::get` apparently wants to create an object of type `T` on the stack, which in this case would be my wrapper object.

Here's a somewhat minimal example of what I'm trying to achieve; this used to compile just fine up to sol 2.20.2:
```
#include <sol.hpp>

class Container
{
public:
	typedef sol::object value_type;

	struct iterator
	{
		typedef size_t size_type;
		typedef ptrdiff_t difference_type;
		typedef sol::object value_type;
		typedef sol::object reference;
		typedef sol::object pointer;
		typedef std::random_access_iterator_tag iterator_category;

		Container &m_container;
		size_type m_index;

		iterator(Container &container, size_type sample) noexcept : m_container(container), m_index(sample) { }
		iterator(const iterator &other, size_type sample) noexcept : m_container(other.m_container), m_index(sample) { }
		iterator(const iterator&) noexcept = default;
		iterator(iterator&&) noexcept = default;

		iterator operator=(const iterator &other) { m_index = other.m_index; }
		friend void swap(iterator &lhs, iterator &rhs) { std::swap(lhs.m_index, rhs.m_index); }

		reference operator*() const { return m_container[m_index]; }
		reference operator[](size_type i) const { return m_container[m_index + i]; }

		iterator operator++() { return iterator(*this, m_index++); }
		iterator operator--() { return iterator(*this, m_index--); }
		iterator operator++(int) { return iterator(*this, ++m_index); }
		iterator operator--(int) { return iterator(*this, --m_index); }

		iterator operator+=(size_type i) { m_index += i; return iterator(*this, m_index); }
		iterator operator-=(size_type i) { m_index -= i; return iterator(*this, m_index); }

		friend iterator operator+(const iterator &lhs, size_type i) { return iterator(lhs, lhs.m_index + i); }
		friend iterator operator-(const iterator &lhs, size_type i) { return iterator(lhs, lhs.m_index + i); }
		friend iterator operator+(size_type i, const iterator &rhs) { return iterator(rhs, rhs.m_index + i); }
		friend iterator operator-(size_type i, const iterator &rhs) { return iterator(rhs, rhs.m_index + i); }
		difference_type operator-(const iterator &rhs) { return m_index - rhs.m_index; }

		bool operator==(const iterator &rhs) const { return m_index == rhs.m_index; }
		bool operator!=(const iterator &rhs) const { return m_index != rhs.m_index; }
		bool operator< (const iterator &rhs) const { return m_index < rhs.m_index; }
		bool operator> (const iterator &rhs) const { return m_index > rhs.m_index; }
		bool operator<=(const iterator &rhs) const { return m_index <= rhs.m_index; }
		bool operator>=(const iterator &rhs) const { return m_index >= rhs.m_index; };
	};

	Container(int foo) { }

	size_t size() { return 5; }
	iterator begin() { return iterator(*this, 0); }
	iterator end() { return iterator(*this, 5); }
	value_type operator[] (size_t index) { return 0; }
};

class Owner
{
	Container container;
public:
	Owner(int foo) : container(foo) { }
	static void Register(sol::table table)
	{
		table.new_usertype<Owner>("owner",
			"container", &Owner::container);
	}
};

int main(int, char*[])
{
	sol::state lua;
	Owner::Register(lua.globals());
}
```

**Comment by ThePhD** [comment]
Sounds like a regression to me. I'll look into it.

**Comment by ThePhD** [comment]
Oh, this is a slight regression because of some work we are doing.

So, previously when you retrieved a container by reference or value we would differentiate the cases. If you ever bound anything that returned by value we would use a converting constructor: by reference, we would simply `stack::get` the usertype's memory.

Right now, a feature was added that allowed you to perform conversions at the function-argument level. That is, `void f( std::vector<int> );` works as a way to convert from a table without having to specifically mark it up `as_table`. (The version with using `std::veector<int> blah = lua["blah"];` still won't work until I beat C++ into shape with a proposal I am writing).

Long story short: the value getter requires a default constructor and that causes issues. I'll put an additional barrier to check for default constructibility to save you some pain here.

**Comment by sagamusix** [comment]
So there are no plans to undo this requirement for containers that are not meant to ever be converted to from Lua code?

**Comment by ThePhD** [comment]
No: I've fixed it and regenerated the single include. Pull the latest: you should be all set.

As a side note, a container without a default constructor might not be All Good™ according to the C++ standard. The current fix is that you just don't get the free conversion to/from Lua code if you container doesn't have a default constructor (e.g., is Sufficiently Special™).

**Comment by sagamusix** [comment]
Fix confirmed. Thanks a lot!
